### ISSUE TITLE
RFE: use learner mode for joining etcd members

### ISSUE THREAD

**Issue by fabriziopandini** [kubernetes/kubeadm#1793]
Growing a local etcd cluster is a complex operation, and in the past, we already faced some issues like e.g. https://github.com/kubernetes-sigs/kind/issues/588

Now that the implementation of the etcd learner mode is progressing, we should start considering if to use it in kubeadm in order to make join --control-plane implementation more robust.

at a high level what we would like to achieve is:
- a new etcd member should be created as a learner and became a voting member only after the etcd data are fully aligned.
ideally
- we should also prevent the api-server to read from a learner node

Ref docs:
- https://etcd.io/docs/v3.3.12/learning/learner/
- https://etcd.io/docs/v3.4.0/learning/design-learner/

------------

(edit by neolit123)

1.26:
- initial KEP draft
https://hackmd.io/@DAKGcrh_RpC5vlt8w5bf8A/r1qoLh9zj
- KEP PR + tracking issue at k/e:
https://github.com/kubernetes/enhancements/pull/3615
https://github.com/kubernetes/enhancements/issues/3614

1.27(alpha):
- k/k alpha feature gate code
https://github.com/kubernetes/kubernetes/pull/113318
https://github.com/kubernetes/kubernetes/pull/115038
- CI added
https://github.com/kubernetes/kubeadm/pull/2807
https://github.com/kubernetes/test-infra/pull/28574

1.29(beta):
- https://github.com/kubernetes/enhancements/pull/4172
- https://github.com/kubernetes/kubernetes/pull/120228
- https://github.com/kubernetes/website/pull/42801
- https://github.com/kubernetes/website/pull/43095

1.32(GA):
- https://github.com/kubernetes/kubernetes/pull/126374
- https://github.com/kubernetes/kubeadm/pull/3096
- https://github.com/kubernetes/test-infra/pull/33316
- https://github.com/kubernetes/website/pull/47563
- https://github.com/kubernetes/kubernetes/pull/127491 and https://github.com/kubernetes/kubernetes/pull/127650 (Cherry-picked to v1.28-1.31)

1.33:
- https://github.com/kubernetes/kubernetes/pull/129589
- https://github.com/kubernetes/website/pull/49410

**Comment by rosti** [comment]
We have to be careful, but we certainly need to act upon it. The plan is that from etcd 3.5 new members will be joined only as learners.
In etcd 3.5 it will be possible to use a learner node for reading, but still the problem with writing continues. And, as LBs are out of the scope of kubeadm, things might become a bit difficult.
We probably need to direct API servers to healthy leaders and possibly do that via an etcd LB. Another possibility is to not expose the API servers, that have a local learner etcd node from the API LB (not sure if this would actually work though).
In short, we need to experiment a bit with this to find what's viable and easy for use.

**Comment by neolit123** [comment]
xref https://github.com/kubernetes/kubeadm/issues/2005#issuecomment-575931176

**Comment by ereslibre** [comment]
Some context on this: https://github.com/etcd-io/etcd/pull/11640, we might want to wait for an etcd version that includes this patch.

**Comment by pacoxu** [comment]
Since Kubernetes 1.23, we are using etcd 3.5.0 now.
https://etcd.io/docs/v3.5/learning/design-learner/
This seems to be a nice candidate for 1.24.

**Comment by pacoxu** [comment]
What is the suggested way to use the learner mode?

1. join member as a learner
2. wait for data aligning and then promote it to a member (kubeadm needs to wait for it to happen.)

And I see in the etcd future plan, https://github.com/etcd-io/etcd/pull/10887 a pull request that may add the ability to auto promote the learner. (Will we wait for the auto promote feature?)

**Comment by neolit123** [comment]
I am not up to date on the learner mode support in etcd but auto promotion
sounds better. The change in kubeadm  will need a new KEP.

https://github.com/kubernetes/enhancements/tree/master/keps

**Comment by neolit123** [comment]
we had a discussion about this today with @fabriziopandini 
so learner mode was used in Talos
https://www.talos.dev/v0.12/introduction/what-is-new/#etcd
and we might want to have it in kubeadm.

i can start writing a doc / KEP for this but if someone will like to work on the code / unit tests / e2e tests that would be great.

**Comment by neolit123** [comment]
@fabriziopandini @pacoxu 
and others watching the ticket.

i've created a hack MD draft for the learner mode proposal:
https://hackmd.io/XZHxeO21T-GDgRhjtVwlGQ

comments are welcome.
we can notify kubespray and Cluster API for feedback too.

**Comment by neolit123** [comment]
once we have some agreement, i can PR the KEP in a raw / draft stage and keep it provisional.
if we get a contributor who wants to work on this we can start tracking the feature for a kubeadm / k8s release milestone.

**Comment by sbueringer** [comment]
Thx! I've added it to the ClusterAPI office hours agenda for today. 

I'll also try to take a look, but could take a bit, just not enough bandwith right now.

**Comment by neolit123** [comment]
updated link https://hackmd.io/@DAKGcrh_RpC5vlt8w5bf8A/r1qoLh9zj

**Comment by pacoxu** [comment]
The KEP looks good. For the detail steps according to etcd docs.
1. add as a learner: `member add --learner`.
2. promote it once `canPromote==true`: call `member promote` API. There would be a new timeout here.

My only question is that:
- Can we use an etcd learner as a standby node? I mean we don't promote it and leave it to users. Users can promote it in case the master has some problems. Or keep it as a learner/standby node for failover. Should we add a flag for it?

**Comment by neolit123** [comment]
> Can we skip step 2 above so that we use an etcd learner as a standby node? I mean we don't promote it and leave it to users. Users can promote it in case the master has some problems. Or keep it as a learner/standby node for failover. Should we add a flag for it?

my vote would be to preserve the current behavior and not a have a flag - i.e. always try to promote.

> BTW, there is no new learner mode related bug in etcd repo.(I am not sure if there is no enough feature users in etcd community or the feature is very stable.)

that may be a concern, but the Talos project is using it too and maybe it just works fine.

we can ask etcd maintainers later.

**Comment by neolit123** [comment]
KEP PR is here:
https://github.com/kubernetes/enhancements/pull/3615

**Comment by juliantaylor** [comment]
How was the issue with the apiserver failing when requesting against a learner node addressed?
https://github.com/kubernetes/kubeadm/issues/1793#issuecomment-532630048
https://github.com/etcd-io/etcd/issues/12789

As of etcd 3.5.6 and kubernetes 1.23.14 write requests will fail when it hits a learner node:

     etcdserver: rpc not supported for learner
     
Is this handled in newer k8s-apiserver versions?

**Comment by tobiasgiese** [comment]
I added a WIP PR with a tested implementation of the learner

**Comment by ahrtr** [comment]
> @juliantaylor TMK, no. also this is the first time i see the issue.

Basically there is no any change on the logic in etcdserver in 3.5,
- [interceptor.go#L53-L55](https://github.com/etcd-io/etcd/blob/715a0047faba060577841b13c87e9b6a1269eaa0/server/etcdserver/api/v3rpc/interceptor.go#L53-L55)
- [interceptor.go#L222-L224](https://github.com/etcd-io/etcd/blob/715a0047faba060577841b13c87e9b6a1269eaa0/server/etcdserver/api/v3rpc/interceptor.go#L222-L224)

So the error message below should `NOT` be a new issue in etcd 3.5.x.
```
As of etcd 3.5.6 and kubernetes 1.23.14 write requests will fail when it hits a learner node:

 etcdserver: rpc not supported for learner

```

**Comment by pacoxu** [comment]
I only tested it in my clusters and maybe the etcd manifest is left before I rejoin. That is why I did not meet this bug.
- TODO: we may add some e2e test case for this mode.

Logically, this would be a problem. Thanks for the fix and I will test and review it later.

**Comment by juliantaylor** [comment]
> > @juliantaylor TMK, no. also this is the first time i see the issue.
> 
> Basically there is no any change on the logic in etcdserver in 3.5,
> 
>     * [interceptor.go#L53-L55](https://github.com/etcd-io/etcd/blob/715a0047faba060577841b13c87e9b6a1269eaa0/server/etcdserver/api/v3rpc/interceptor.go#L53-L55)
> 
>     * [interceptor.go#L222-L224](https://github.com/etcd-io/etcd/blob/715a0047faba060577841b13c87e9b6a1269eaa0/server/etcdserver/api/v3rpc/interceptor.go#L222-L224)
> 
> 
> So the error message below should `NOT` be a new issue in etcd 3.5.x.
> 
> ```
> As of etcd 3.5.6 and kubernetes 1.23.14 write requests will fail when it hits a learner node:
> 
>  etcdserver: rpc not supported for learner
> ```

yes it is not a new problem it exists since the learner mode exists.
It does imply you need to reconfigure and restart the apiservers every time _before_ an etcd instance is restarted, if that is handled by kubeadm there is no issue.

**Comment by juliantaylor** [comment]
ah restarting is not a problem as it does not put the node back into learning mode. So from kubeadm's perspective this everything should be ok if it adds the node to the apiserver after promotion.

**Comment by ahrtr** [comment]
FYI. https://github.com/etcd-io/etcd/issues/15107

**Comment by chendave** [comment]
/remove-lifecycle stale

Looks like this has been implemented, but should stay open until this is graduated.

**Comment by pacoxu** [comment]
Updated the issue description for the implementation history:

The CI https://testgrid.k8s.io/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-learner-mode-latest. This feature is alpha since 1.27. I wondered if we could promote it to beta in the v1.29 release cycle.

Is there anything further that we should do to promote it? I opened https://github.com/kubernetes/kubernetes/pull/120228.

**Comment by chrischdi** [comment]
> Updated the issue description for the implementation history:
> 
> The CI https://testgrid.k8s.io/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-learner-mode-latest. This feature is alpha since 1.27. I wondered if we could promote it to beta in the v1.29 release cycle.
> 
> Is there anything further that we should do to promote it? I opened [kubernetes/kubernetes#120228](https://github.com/kubernetes/kubernetes/pull/120228).

@tobiasgiese : maybe you could provide some feedback, I think you folks still use it right?

**Comment by tobiasgiese** [comment]
> maybe you could provide some feedback, I think you folks still use it right?

We are using it already, yes. Also we have backported it to v1.2[4-6] (since https://github.com/kubernetes/kubernetes/pull/115038) and it is working quite well.
We have never had any problems with the learner mode and we have alot of nightly builds (about 50 periodic nightly builds and 40 Prow trigger builds/jobs).

**Comment by pacoxu** [comment]
I suppose that we should graduate this feature later in v1.31+ and get more feedback before GA.

**Comment by neolit123** [comment]
> I suppose that we should graduate this feature later in v1.31+ and get more feedback before GA.
> 
> So no action item for v1.30.

i got contacted in slack by a person that had feedback about learner mode in kubeadm, but they never send me the info.
learner mode was broken for them in some way.

i will see if i can message them about this after NY.

**Comment by pacoxu** [comment]
One issue that I may imagine is timeout for a step of promotion ready waiting or promotion may be a problem.

**Comment by pacoxu** [comment]
https://github.com/kubernetes/kubeadm/issues/2997#issuecomment-1899856003
We have a short discussion about if we need to add progress percentage of synced in logging. 
- It needs change in etcd side. 

https://github.com/kubernetes/kubeadm/issues/2997#issuecomment-1899805047
Another potentiel improvement is adding a configurable timeout for etcd learner ready for promoting. There are already a lot of timeout configuration in v1beta4 timeouts structs. (+0 for this as 2 min should be enough for most scanerios.)

**Comment by neolit123** [comment]
> We have a short discussion about if we need to add progress percentage of synced in logging.

ok, i don't think it's GA blocking.

> Another potentiel improvement is adding a configurable timeout for etcd learner ready for promoting. There are already a lot of timeout configuration in v1beta4 timeouts structs. (+0 for this as 2 min should be enough for most scanerios.)

+0 as well from me.
our 2 minutes timeout will apply to all etcd client calls by default.

**Comment by neolit123** [comment]
> i got contacted in slack by a person that had feedback about learner mode in kubeadm, but they never send me the info.
learner mode was broken for them in some way.

they did not log an issue...

**Comment by pacoxu** [comment]
I updated beta related PRs in this issue description.

I think we may wait for at least another 2 release cycles for feedbacks to make this GA, as most users are not using v1.29 yet, which make it beta, by default enabled.

**Comment by pacoxu** [comment]
https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-kubeadm-kinder-upgrade-addons-before-controlplane-1-29-latest/1777803167627481088

A flake:
```
I0409 21:02:09.391263     303 local.go:165] Updated etcd member list: [{kinder-upgrade-addons-before-controlplane-control-plane-2 https://172.17.0.3:2380/} {kinder-upgrade-addons-before-controlplane-control-plane-1 https://172.17.0.2:2380/}]
I0409 21:02:09.428719     303 etcd.go:508] [etcd] Promoting a learner as a voting member: a8f2efe87cf50990
{"level":"warn","ts":"2024-04-09T21:02:09.438938Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0007de1c0/172.17.0.2:2379","attempt":0,"error":"rpc error: code = FailedPrecondition desc = etcdserver: can only promote a learner member which is in sync with leader"}
I0409 21:02:09.439065     303 etcd.go:533] [etcd] Promoting the learner a8f2efe87cf50990 failed: etcdserver: can only promote a learner member which is in sync with leader
{"level":"warn","ts":"2024-04-09T21:02:09.547427Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0007de1c0/172.17.0.2:2379","attempt":0,"error":"rpc error: code = FailedPrecondition desc = etcdserver: can only promote a learner member which is in sync with leader"}
I0409 21:02:09.547508     303 etcd.go:533] [etcd] Promoting the learner a8f2efe87cf50990 failed: etcdserver: can only promote a learner member which is in sync with leader
{"level":"warn","ts":"2024-04-09T21:02:09.700675Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0007de1c0/172.17.0.2:2379","attempt":0,"error":"rpc error: code = FailedPrecondition desc = etcdserver: can only promote a learner member which is in sync with leader"}
I0409 21:02:09.700752     303 etcd.go:533] [etcd] Promoting the learner a8f2efe87cf50990 failed: etcdserver: can only promote a learner member which is in sync with leader
{"level":"warn","ts":"2024-04-09T21:02:09.942281Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0007de1c0/172.17.0.2:2379","attempt":0,"error":"rpc error: code = FailedPrecondition desc = etcdserver: can only promote a learner member which is in sync with leader"}
I0409 21:02:09.942448     303 etcd.go:533] [etcd] Promoting the learner a8f2efe87cf50990 failed: etcdserver: can only promote a learner member which is in sync with leader
{"level":"warn","ts":"2024-04-09T21:02:10.307133Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0007de1c0/172.17.0.2:2379","attempt":0,"error":"rpc error: code = FailedPrecondition desc = etcdserver: can only promote a learner member which is in sync with leader"}
I0409 21:02:10.307218     303 etcd.go:533] [etcd] Promoting the learner a8f2efe87cf50990 failed: etcdserver: can only promote a learner member which is in sync with leader
{"level":"warn","ts":"2024-04-09T21:02:10.832218Z","logger":"etcd-client","caller":"v3/retry_interceptor.go:62","msg":"retrying of unary invoker failed","target":"etcd-endpoints://0xc0007de1c0/172.17.0.2:2379","attempt":0,"error":"rpc error: code = FailedPrecondition desc = etcdserver: can only promote a learner member which is in sync with leader"}
I0409 21:02:10.832305     303 etcd.go:533] [etcd] Promoting the learner a8f2efe87cf50990 failed: etcdserver: can only promote a learner member which is in sync with leader
I0409 21:02:11.660333     303 etcd.go:530] [etcd] The learner was promoted as a voting member: a8f2efe87cf50990
[etcd] Waiting for the new etcd member to join the cluster. This can take up to 40s
```

40s timeout for waiting to be ready to promote a learner.

**Comment by neolit123** [comment]
> 40s timeout for waiting to be ready to promote a learner.

should we increase this time to 2 minutes, or more by default?

**Comment by pacoxu** [comment]
https://github.com/kubernetes/kubernetes/blob/227c2e7c2b2c05a9c8b2885460e28e4da25cf558/cmd/kubeadm/app/util/etcd/etcd.go#L531-L557

already 2m.

I miss the log that `The learner was promoted as a voting member` success finally. Sorry for disturb.

**Comment by neolit123** [comment]
the flakes on https://prow.k8s.io/job-history/gs/kubernetes-jenkins/logs/ci-kubernetes-e2e-kubeadm-kinder-upgrade-addons-before-controlplane-1-29-latest
 seem like slow infra problems, 5 minutes should be plenty of time for a few nodes to join and be ready :/

**Comment by neolit123** [comment]
> I think we may wait for at least another 1 or 2 release cycles for feedbacks to make this GA, as most users are not using v1.29 yet, which make it beta, by default enabled.

@pacoxu should we GA this in 1.32?

**Comment by pacoxu** [comment]
Agree
